pkg/tls/tls_test.go

```diff
@@ -0,0 +1,158 @@
+	}
+	knownGoodDER, err := x509.CreateCertificate(rand.Reader, template, template, pk.Public(), pk)
+	if err != nil {
+		fmt.Errorf("Error generating known good cert %s", err)
+	}
+	knownGood, err := x509.ParseCertificate(knownGoodDER)
+	if err != nil {
+		fmt.Errorf("unable to parse DER %s", err)
+	}
+	return knownGood
+}
+
+var cn = "dummy-test-certificate.local"
+
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	if newPK.Validate() != nil {
+		t.Errorf("Failed to validate private key")
+	}
+	// the exponent which is effectively the length is encoded inside PublicKey
+	keyLength := newPK.PublicKey.N.BitLen()
+	if keyLength < 2048 {
+		t.Errorf("insecure key length detected: %d", keyLength)
+	}
+}
+
+func TestNewCA(t *testing.T) {
+	pk, cert, err := NewCA(cn)
+	if err != nil {
+		t.Errorf("failed to create new CA %s", err)
+	}
+	if pk.Validate() != nil {
+		t.Errorf("failed to validate RSA Private key")
+	}
+	//opts := x509.VerifyOptions{DNSName: cn}
+	//if _, err := cert.Verify(opts); err != nil {
+	//	t.Errorf("Failed to verify created self signed certificate %s", err)
+	//}
+	fmt.Printf("%#v\n", cert)
+}
+
+func TestNewSelfSignedCert(t *testing.T) {
+	const knownGoodLoc = "./testdata/known_good_cert.pem"
+	cn := cn
+	org := []string{"Red Hat Internal"}
+	dnsNames := []string{cn}
+	ipAddresses := []net.IP{net.ParseIP("192.168.0.1")}
+	keyUsage := []x509.ExtKeyUsage{x509.ExtKeyUsageAny}
+	selfSign := true
+
+	pk, cert, err := NewCert(cn, org, dnsNames, ipAddresses, keyUsage, nil, nil, selfSign)
+	if err != nil {
+		t.Errorf("error creating new cert %s", err)
+	}
+	if pk.Validate() != nil {
+		t.Errorf("error validating private key %s", err)
+	}
+	//opts := x509.VerifyOptions{DNSName: cn}
+	//if _, err := cert.Verify(opts); err != nil {
+	//	t.Errorf("cannot verify certificate %s", err)
+	//}
+	const updateSSLVar = "UPDATE_KNOWN_SSL_CERT"
+	if os.Getenv(updateSSLVar) == "true" {
+		// reuse functionality from marshal.go
+		pemEncodedCert, err := CertAsBytes(getKnownGoodCert())
+		if err != nil {
+			t.Errorf("Cannot PEM encode certificate")
+		}
+		err = ioutil.WriteFile(knownGoodLoc, pemEncodedCert, 0644)
+		if err != nil {
+			t.Errorf("unable to save PEM cert %s", err)
+		}
+	}
+	// we have a cert, let's load the last good one from disk
+	knownGoodBlob, err := ioutil.ReadFile(knownGoodLoc)
+	if err != nil {
+		t.Errorf("unable to load last known good cert from disk %s", err)
+	}
+	block, _ := pem.Decode(knownGoodBlob)
+	if block == nil || block.Type != "CERTIFICATE" {
+		t.Errorf("failed to decode PEM block containing public key")
+	}
+
+	knownGoodCerts, err := x509.ParseCertificates(block.Bytes)
+	if err != nil {
+		t.Errorf("unable to parse certificates %s", err)
+	}
+	goodCert := knownGoodCerts[0]
+	cert.SerialNumber = big.NewInt(0)
+	cert.NotBefore = time.Unix(0, 0)
+	cert.NotAfter = time.Unix(0, 0)
+	if reflect.DeepEqual(goodCert, cert) == false {
```

Comment:
`!reflect.DeepEqual(goodCert,cert)` or `reflect.DeepEqual(goodCert,cert)` depending on what you want.